File: clients/haskell/src/Debugging/DebugState.hs
module Debugging.DebugState where

import Prelude hiding (id)
import qualified Trans
import Trans (Trans)
import Data.Int
import Debugging.Camera (Camera)
import Model.Vec2 (Vec2)

-- | Renderer's state
data DebugState = DebugState {
    -- | Pressed keys
    pressedKeys :: [String],
    -- | Cursor's position in game coordinates
    cursorWorldPosition :: Vec2,
    -- | Id of unit which is followed by the camera, or None
    lockedUnit :: Maybe Int32,
    -- | Current camera state
    camera :: Camera }
    deriving Show

instance Trans DebugState where
    read = do
        pressedKeys <- Trans.read
        cursorWorldPosition <- Trans.read
        lockedUnit <- Trans.read
        camera <- Trans.read
        return DebugState {
            pressedKeys,
            cursorWorldPosition,
            lockedUnit,
            camera }
    
    write DebugState {
        pressedKeys,
        cursorWorldPosition,
        lockedUnit,
        camera } = do
            Trans.write pressedKeys
            Trans.write cursorWorldPosition
            Trans.write lockedUnit
            Trans.write camera

File: clients/rust/src/debugging/debug_state.rs
use super::*;

/// Renderer's state
#[derive(Clone, Debug)]
pub struct DebugState {
    /// Pressed keys
    pub pressed_keys: Vec<String>,
    /// Cursor's position in game coordinates
    pub cursor_world_position: model::Vec2,
    /// Id of unit which is followed by the camera, or None
    pub locked_unit: Option<i32>,
    /// Current camera state
    pub camera: debugging::Camera,
}

impl trans::Trans for DebugState {
    fn write_to(&self, writer: &mut dyn std::io::Write) -> std::io::Result<()> {
        self.pressed_keys.write_to(writer)?;
        self.cursor_world_position.write_to(writer)?;
        self.locked_unit.write_to(writer)?;
        self.camera.write_to(writer)?;
        Ok(())
    }
    fn read_from(reader: &mut dyn std::io::Read) -> std::io::Result<Self> {
        let pressed_keys: Vec<String> = trans::Trans::read_from(reader)?;
        let cursor_world_position: model::Vec2 = trans::Trans::read_from(reader)?;
        let locked_unit: Option<i32> = trans::Trans::read_from(reader)?;
        let camera: debugging::Camera = trans::Trans::read_from(reader)?;
        Ok(Self {
            pressed_keys,
            cursor_world_position,
            locked_unit,
            camera,
        })
    }
}